clash-cores/src/Clash/Cores/Xilinx/DcFifo/Explicit.hs

```diff
@@ -0,0 +1,254 @@
+    ( Signal write Full
+    , Signal write Bool
+    , Signal write (DataCount depth)
+
+    , Signal read Empty
+    , Signal read Bool
+    , Signal read (DataCount depth)
+    , Signal read (BitVector n)
+    )
+  go st@(FifoState _ rt) rstR rEna rstW =
+    if rt < tWr
+      then goRead st rstR rEna rstW
+      else goWrite st rstR rEna rstW
+
+  goWrite (FifoState _ rt) rstR rEna (True :- rstWNext) (_ :- wData) =
+      (False :- preFull, False :- preOver, 0 :- preWCnt, fifoEmpty, under, rCnt, rData)
+    where
+      (preFull, preOver, preWCnt, fifoEmpty, under, rCnt, rData) =
+        go (FifoState mempty (rt-tWr)) rstR rEna rstWNext wData
+
+  goWrite (FifoState q rt) rstR rEna (_ :- rstW) (wDat :- wDats1) =
+    (full, over, wCnt, fifoEmpty, under, rCnt, rData)
+    where
+      (preFull, preOver, preWCnt, fifoEmpty, under, rCnt, rData) =
+        go (FifoState q' (rt-tWr)) rstR rEna rstW wDats1
+
+      wCnt = sDepth q :- preWCnt
+      full = (Seq.length q == maxDepth) :- preFull
+      (q', over) =
+        if Seq.length q < maxDepth
+          then (case wDat of { Just x -> x Seq.<| q ; _ -> q }, False :- preOver)
+          else (q, True :- preOver)
+
+  sDepth = fromIntegral . Seq.length
+
+  goRead (FifoState _ rt) (True :- rstRNext) (_ :- rEnas1) rstW wData =
+    (full, over, wCnt, fifoEmpty, under, rCnt, rData)
+    where
+      rData = deepErrorX "Reset" :- preRData
+      fifoEmpty = True :- preEmpty
+      rCnt = 0 :- preRCnt
+      under = False :- preUnder
+
+      (full, over, wCnt, preEmpty, preUnder, preRCnt, preRData) =
+        go (FifoState mempty (rt+tR)) rstRNext rEnas1 rstW wData
+
+  goRead (FifoState q rt) (_ :- rstRNext) (rEna :- rEnas1) rstW wData =
+    (full, over, wCnt, fifoEmpty, under, rCnt, rData)
+    where
+      rCnt = sDepth q :- preRCnt
+      fifoEmpty = (Seq.length q == 0) :- preEmpty
+      rData = nextData :- preRData
+
+      (full, over, wCnt, preEmpty, preUnder, preRCnt, preRData) =
+        go (FifoState q' (rt+tR)) rstRNext rEnas1 rstW wData
+
+      (q', nextData, under) =
+        if rEna
+          then
+            case Seq.viewr q of
+              Seq.EmptyR -> (q, deepErrorX "FIFO empty", True :- preUnder)
+              qData Seq.:> qDatum -> (qData, qDatum, False :- preUnder)
+          else (q, deepErrorX "Enable off", deepErrorX "Enable off" :- preUnder)
+
+  initState :: FifoState n
+  initState = FifoState Seq.empty 0
+
+  tWr = snatToNum @Int (clockPeriod @write)
+  tR = snatToNum @Int (clockPeriod @read)
+
+{-# NOINLINE dcFifo #-}
+{-# ANN dcFifo (InlineYamlPrimitive [minBound..maxBound] [i|
+BlackBoxHaskell:
+    name: Clash.Cores.Xilinx.DcFifo.Explicit.dcFifo
+    templateFunction: Clash.Cores.Xilinx.DcFifo.BlackBoxes.dcFifoBBF
+    workInfo: Always
+    #    multiResult: true
+    includes:
+    - name: fifo
+      extension: tcl
+      format: Haskell
+      templateFunction: Clash.Cores.Xilinx.DcFifo.BlackBoxes.dcFifoTclTF
```

Comment:
The include template is defined both here in YAML and in `Clash.Cores.Xilinx.DcFifo.BlackBoxes.dcFifoBBF`. Is this necessary? Since usually the type of the `templateFunction` in the YAML is `TemplateFunction`, and that is not the type of `dcFifoTclTF`, it seems the YAML is not used, as there is no error about the type.

Reply:
Yep!